src/PasswordHistory.php

```diff
@@ -0,0 +1,207 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ * ---------------------------------------------------------------------
+ */
+
+ /**
+  * Helper class that contains method to handle password history validation or
+  * change
+  *
+  * Implemented temporarily as a singleton, should become some kind of service
+  * once GLPI support depencency injection.
+  */
+final class PasswordHistory
+{
+    /**
+     * Max allowed history size
+     */
+    public const MAX_HISTORY_SIZE = 4;
+
+    /**
+     * Singleton instance
+     */
+    private static ?PasswordHistory $instance = null;
+
+    /**
+     * Get singleton instance
+     *
+     * @return PasswordHistory
+     */
+    public static function getInstance(): self
+    {
+        if (!self::$instance) {
+            self::$instance = new self();
+        }
+
+        return self::$instance;
+    }
+
+    /**
+     * Validate that the user password was not used previously
+     *
+     * @param User $user
+     * @param string $password
+     *
+     * @return bool
+     */
+    public function validatePassword(
+        User $user,
+        string $password
+    ): bool {
+        $length = $this->getConfiguratedLength();
+
+        // Check if user is being created
+        if ($user->isNewItem()) {
+            // No history to handle for a new user, no need to go further
+            return true;
+        }
+
+        // Check current password
+        if (Auth::checkPassword($password, $user->fields['password'])) {
+            return false;
+        }
+
+        // Check if password history is enabled
+        if ($length == 0) {
+            // No password history, no need to go further
+            return true;
+        }
+
+        $passwords_checked_count = 0;
+        foreach ($this->getForUser($user) as $previous_password) {
+            if (Auth::checkPassword($password, $previous_password)) {
+                // New password match an entry in the history
+                return false;
+            }
+
+            // History might contain more entries than configurated (skip them)
+            $passwords_checked_count++;
+            if ($passwords_checked_count == $length) {
+                break;
+            }
+        }
+
+        // New password didn't match any entry in the history
+        return true;
+    }
+
+    /**
+     * Update password history for a given user
+     * Must be called after a user password is updated
+     *
+     * @param User $user
+     *
+     * @return void
+     */
+    public function updatePasswordHistory(User $user): void
+    {
+        $password = $user->oldvalues['password'] ?? '';
+
+        // No password change
+        if (empty($password)) {
+            return;
+        }
```

Comment:
`$password` should be passed to the method. Using value from `$user->oldvalues` makes this method not usable outside the `post_updateItem` context.

Reply:
Indeed, good catch.
Done in 98d1d2c406.